### PR TITLE
libclc: remove __attribute__((assume)) for clspv targets

### PR DESCRIPTION
**Pull request by rjodinchr**
Instead add a proper attribute in clang, and add convert it to function metadata to keep the information in the IR. The goal is to remove the dependency on __attribute__((assume)) that should have not be there in the first place.

Ref https://github.com/llvm/llvm-project/pull/84934

### CODE REVIEW

File: clang/include/clang/Basic/Attr.td
Comment (erichkeane) [inline clang/include/clang/Basic/Attr.td:4567]: Please add documentation anyway.  See `AttrDocs.td`

File: clang/include/clang/Basic/Attr.td
Comment (erichkeane) [inline clang/include/clang/Basic/Attr.td:4568]: This should also probably set 'appertainment' via `Subjects`.

**Review by erichkeane — COMMENTED** [review]
Also needs:
-a release note
-A test: At least a test against LLVM-IR (CodeGen test).

File: clang/include/clang/Basic/Attr.td
Comment (Sirraide) [reply clang/include/clang/Basic/Attr.td:4568]: Also, is there a reason as to why it’s a `DeclOrStmtAttr` instead of an `InheritableAttr`? Because afaik most of the other function attributes use the latter.

Comment (rjodinchr) [reply clang/include/clang/Basic/Attr.td:4568]: I have never dealt with clang attribute implementation, so I have no idea whether `DeclOrStmtAttr` or `InheritableAttr` should be used. I'll change it for `InheritableAttr`.

File: clang/include/clang/Basic/Attr.td
Comment (rjodinchr) [reply clang/include/clang/Basic/Attr.td:4567]: What should be the documentation category?

Comment (erichkeane) [reply clang/include/clang/Basic/Attr.td:4567]: `DocCatFunction`?  As it seems to be a function decl attribute.  Subjects should be functions only it appears.

**Comment by rjodinchr** [comment]
@erichkeane, I don't know what you expect for the release note, where should I write something?

**Comment by erichkeane** [comment]
> @erichkeane, I don't know what you expect for the release note, where should I write something?

See `ReleaseNotes.rst`, find a section that makes sense (probably attribute changes), and write a quick blurb about the attribute, likely a very condensed version of the docs.

File: clang/docs/ReleaseNotes.rst
Comment (erichkeane) [inline clang/docs/ReleaseNotes.rst:396]:
```suggestion
- The ``clspv_libclc_builtin`` attribute has been added to allow clspv to identify functions coming from libclc.
```

I'd suggest clarifying via a parenthetical what clspv and libclc are.

File: clang/include/clang/Basic/AttrDocs.td
Comment (erichkeane) [inline clang/include/clang/Basic/AttrDocs.td:8094]: I think this needs quite a bit more details, please see the rest of the AttrDocs examples, they typically go into examples, and at least specify what things like clspv and libclc would mean.

**Review by erichkeane — APPROVED** [review]

**Comment by Sirraide** [comment]
There’s a test that checks all attributes that we support (Misc/pragma-attribute-supported-attributes-list.test) and which is failing at the moment; you’ll have to add this attribute there as well.

**Review by erichkeane — APPROVED** [review]
Still fine once CI approves

**Comment by rjodinchr** [comment]
Is there something else needed on my side to get that PR merged?

**Comment by erichkeane** [comment]
> Is there something else needed on my side to get that PR merged?

My apologies, I thought you'd been around LLVM enough that you had merge rights.  I've done it now.

**Pull request merged**
